### ISSUE TITLE
Not calibrating on Raspberry Pi with 800x320 display

### ISSUE THREAD

**Issue by richteel** [tom-2015/rpi-AR1100#11]
I successfully ran this software once and have not been able to get it to run again. I did update the firmware on the AR1100 with the Microchip hex files, included in the AR1100 Firmware Download Utility, dated 19 May 2011. I'm not certain but the one time it worked may have been before updating the firmware.

Steps to replicate:
1. Execute the following command from a terminal window on the Raspberry Pi Zero W
$ sudo ./AR1100 -c 9

2.  After a few seconds, the calibration screen is shown with the first calibration point in the upper left corner as expected. Tapping on the screen does not register. The screen remains showing the first calibration point.

3. Using my mouse, I click on the screen to get the application to close. The screen is no longer responding to touch as it is still in generic HID mode.

4.  Execute the following command to return the touch panel to mouse mode.
$ sudo ./AR1100 -m

I recompiled the code with the debug option to see if it provided any hints. Below is the output.

pi@hp95lx-pi:~/rpi-AR1100 $ sudo ./AR1100 -c 9
Unable to find the device.
Connected to AR1100 as USB mouse.
Switching to generic HID mode.
The AR1100 is not in generic HID mode, switching mode.
Sending cmd 0x70
USB tx 64:
55 01 70 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
Unable to find the device.
Unable to find the device.
Unable to find the device.
...Sending cmd 0x13
USB tx 64:
55 01 13 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 13 0c 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

COMMAND response: 0
Calibrating 9 points.
Sending cmd 0x14
USB tx 64:
55 02 14 02 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 14 0c 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

COMMAND response: 0
Screen width is 800, screen height is 320
Calibration point 1: (100,40)
Calibration point 2: (400,40)
Calibration point 3: (700,40)
Calibration point 4: (100,160)
Calibration point 5: (400,160)
Calibration point 6: (700,160)
Calibration point 7: (100,280)
Calibration point 8: (400,280)
Calibration point 9: (700,280)
Drawing point 1: (100,40)
Drawing point 1: (100,40)
Drawing point 1: (100,40)
Received a mouse click, AR1100 not in calibration mode?, exiting 1
The window was forced to close.
Sending cmd 0x13
USB tx 64:
55 01 13 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 14 0c 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

COMMAND response: 0
Setting AR1100 to mouse mode.
Sending cmd 0x71
USB tx 64:
55 01 71 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0

**Comment by richteel** [comment]
BTW: Here is the result after running
$ sudo ./AR1100 -m

pi@hp95lx-pi:~/rpi-AR1100 $ sudo ./AR1100 -m
Connected to AR1100 as generic HID.
Sending cmd 0x13
USB tx 64:
55 01 13 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 14 0c 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

COMMAND response: 0
Error disabeling touch!
Setting AR1100 to mouse mode.
Sending cmd 0x71
USB tx 64:
55 01 71 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0

**Comment by richteel** [comment]
I just noticed that the 5th byte received is 0c. I would expect it to be 00. Could that be an issue?

**Comment by tom-2015** [comment]
It could be the firmware update but as I can see it fails to disable mouse and start calibration because the response is 55 02 00 14 but expected 55 02 00 13.

Can you see what happens when you change:

`if (response.sync==AR1100_SYNC && response.cmd==cmd){`

to

`if (response.sync==AR1100_SYNC){`

in AR1100.cpp:153 and then recompile.

**Comment by richteel** [comment]
No success, same behavior but I did see something different in the receive data that does not look right.

pi@hp95lx-pi:~/rpi-AR1100 $ sudo ./AR1100 -c 9
Unable to find the device.
Connected to AR1100 as USB mouse.
Switching to generic HID mode.
The AR1100 is not in generic HID mode, switching mode.
Sending cmd 0x70
USB tx 64:
55 01 70 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
Unable to find the device.
Unable to find the device.
Unable to find the device.
...Sending cmd 0x13
USB tx 64:
55 01 13 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 13 38 61 20 4c 37 0d 00 dc 00 5c 40 00 
88 41 08 15 40 50 d0 00 b4 40 80 a0 55 84 01 12 
05 22 c4 04 01 c2 32 0a 25 18 59 00 33 85 a1 28 
10 80 18 46 99 04 62 08 04 80 20 40 92 21 04 b1 

COMMAND response: 0
Calibrating 9 points.
Sending cmd 0x14
USB tx 64:
55 02 14 02 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 14 38 61 20 4c 37 0d 00 dc 00 5c 40 00 
88 41 08 15 40 50 d0 00 b4 40 80 a0 55 84 01 12 
05 22 c4 04 01 c2 32 0a 25 18 59 00 33 85 a1 28 
10 80 18 46 99 04 62 08 04 80 20 40 92 21 04 b1 

COMMAND response: 0
Screen width is 800, screen height is 320
Calibration point 1: (100,40)
Calibration point 2: (400,40)
Calibration point 3: (700,40)
Calibration point 4: (100,160)
Calibration point 5: (400,160)
Calibration point 6: (700,160)
Calibration point 7: (100,280)
Calibration point 8: (400,280)
Calibration point 9: (700,280)
Drawing point 1: (100,40)
Drawing point 1: (100,40)
Drawing point 1: (100,40)
Drawing point 1: (100,40)
Received a mouse click, AR1100 not in calibration mode?, exiting 1
The window was forced to close.
Sending cmd 0x13
USB tx 64:
55 01 13 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 14 38 61 20 4c 37 0d 00 dc 00 5c 40 00 
88 41 08 15 40 50 d0 00 b4 40 80 a0 55 84 01 12 
05 22 c4 04 01 c2 32 0a 25 18 59 00 33 85 a1 28 
10 80 18 46 99 04 62 08 04 80 20 40 92 21 04 b1 

COMMAND response: 0
Setting AR1100 to mouse mode.
Sending cmd 0x71
USB tx 64:
55 01 71 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
pi@hp95lx-pi:~/rpi-AR1100 $ sudo ./AR1100 -m
Connected to AR1100 as generic HID.
Sending cmd 0x13
USB tx 64:
55 01 13 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
USB rx 64:
55 02 00 14 38 61 20 4c 37 0d 00 dc 00 5c 40 00 
88 41 08 15 40 50 d0 00 b4 40 80 a0 55 84 01 12 
05 22 c4 04 01 c2 32 0a 25 18 59 00 33 85 a1 28 
10 80 18 46 99 04 62 08 04 80 20 40 92 21 04 b1 

COMMAND response: 0
Setting AR1100 to mouse mode.
Sending cmd 0x71
USB tx 64:
55 01 71 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 

Result 0
pi@hp95lx-pi:~/rpi-AR1100 $

**Comment by richteel** [comment]
I was looking through the code and noticed that the lines "Drawing point 1: (100,40)" should not be displayed 4 times or at least, I cannot determine why it would be executed 4 times.

I also looked at the datasheet as I was wondering why cmd 0x13 is sent at the end but 0x14 is received. This is actually correct as it is in calibration mode. Every time the screen is touched, the response to cmd 0x14 is sent. I'm going to look through the code again tomorrow and see if I can't figure out what is going on. It looks like there may be an issue with the code detecting the presses. Not certain if it is due to a change in how the firmware behaves. It could be that I'm just overlooking something obvious but I believe the firmware is working as per the current datasheet. It could also be with the deprecated code being used as well. Perhaps there is a change in the way some of the legacy code is being executed. I swear that I saw this software seem to work once but maybe I was mistaken.

**Comment by tom-2015** [comment]
Been a long time since I've actually used it. Must be from before Raspberry released it's own touch screen so I forgot a little bit how the AR1100 works. I think I copied the code from the Windows calibration program that Microchip released so you could take a look how it works there.

**Comment by richteel** [comment]
Thanks, I actually have another AR1100 here that I have not touched. I'm going to start with that to see if the application will work with it. If it does, I will be more certain that it was updating the firmware that caused my issues. If it does not work, then I will start changing the code to see if I can get it to work.

I'm glad you have this available as Windows would not support the odd resolution of my display so I could not calibrate it on Windows. If I get this to work, I may make a pull request so others can benefit from it. Thanks for making this available to the community.

**Comment by richteel** [comment]
Same result with one that was not flashed. Maybe I imagined that it worked at one point. On to looking at the software.

**Comment by richteel** [comment]
Think I'm zeroing in on the issue. It looks like it is a timing issue on line 185. I'm trying to see what value for timeout will work but 5 seems too small. It could be because I'm on a Raspberry Pi Zero W.

I also found a bug on the calibration abort, which was not putting the AR1100 back into mouse mode. I have fixed that issue. When I get this timing issue sorted, I will make a note of the changes.

**Comment by richteel** [comment]
Got it working now.

AR1100.cpp
- Change line 185
	FROM: int result = read_data((unsigned char *) & response, sizeof(response), & bytes_read, 5);
	TO: int result = read_data((unsigned char *) & response, sizeof(response), & bytes_read, 50);

The bug that I mentioned is not a bug after all, or at least anything that needs to be addressed. When there was a timeout issue, the mode would not return to previous mode. I had fixed that by issuing the TOUCH_ENABLE command to abort calibration. It appears that is only necessary if the timeout is too short on line 185 of AR1100.cpp.

Tom, thank you so much for putting this application together and responding to my question. I was able to calibrate the screen and now I can access the full desktop on my screen. Before the scaling was off and I could not access the Task Bar items. I can now access the whole desktop.

The only issue I have with the application is that if you dwell too long on a point, the next point is captured as the previous point so your calibration will be off. I thought that perhaps introducing a delay in the application would help but after some thought, I realized that would not help as the AR1100 would need to be the one not to take it as the next point. Without the ability to alter the firmware, that would not be possible. A nice to have would be a way to abort the calibration from the keyboard rather than just a mouse click. I may add that in as there may not always be a mouse connected when running the calibration.

@tom-2015, I'll leave this issue open in case you want to make that one change to the timeout on line 185. Feel free to close this out as it is now resolved. Hopefully this will help someone in the future even if you do not modify that one line. At least this ticket will let them know what to do to get it working in their situation.
